File: backend/src/services/granite.service.ts
import axios, { AxiosError } from "axios";
import dotenv from "dotenv";

dotenv.config();

interface GraniteTokenResponse {
  access_token: string;
  expires_in: number;
}

interface GraniteGenerationResponse {
  model_id?: string;
  created_at?: string;
  results?: Array<{
    generated_text?: string;
    stop_reason?: string;
    generated_token_count?: number;
    input_token_count?: number;
  }>;
}

export interface GraniteResponse {
  success: boolean;
  data: {
    model: string;
    text: string;
    stopReason: string;
    tokens: {
      input: number;
      output: number;
    };
  };
  meta: {
    timestamp: string;
  };
}

let cachedToken: string | null = null;
let tokenExpiresAt = 0;

function getRequiredEnv(name: string): string {
  const value = process.env[name];

  if (!value || !value.trim()) {
    throw new Error(`Missing required environment variable: ${name}`);
  }

  return value.trim();
}

async function getIamToken(): Promise<string> {
  const now = Date.now();

  if (cachedToken && now < tokenExpiresAt) {
    return cachedToken;
  }

  const apiKey = getRequiredEnv("IBM_CLOUD_API_KEY");
  const iamUrl = process.env.IBM_IAM_URL?.trim() || "https://iam.cloud.ibm.com/identity/token";

  const body = new URLSearchParams({
    grant_type: "urn:ibm:params:oauth:grant-type:apikey",
    apikey: apiKey,
  });

  const response = await axios.post<GraniteTokenResponse>(iamUrl, body.toString(), {
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Accept: "application/json",
    },
    timeout: 15000,
  });

  cachedToken = response.data.access_token;
  const expiresInSeconds = response.data.expires_in || 3600;
  tokenExpiresAt = now + Math.max(60, expiresInSeconds - 120) * 1000;

  return cachedToken;
}

function buildGranitePayload(prompt: string) {
  return {
    model_id: process.env.IBM_GRANITE_MODEL_ID?.trim() || "ibm/granite-3-8b-instruct",
    input: prompt,
    parameters: {
      decoding_method: "greedy",
      max_new_tokens: 300,
      min_new_tokens: 1,
      repetition_penalty: 1.05,
    },
    project_id: getRequiredEnv("IBM_WATSONX_PROJECT_ID"),
  };
}

export async function callGranite(prompt: string): Promise<GraniteResponse> {
  if (!prompt || !prompt.trim()) {
    throw new Error("Prompt is required.");
  }

  try {
    const token = await getIamToken();
    const apiUrl =
      process.env.IBM_GRANITE_API_URL?.trim() ||
      "https://us-south.ml.cloud.ibm.com/ml/v1/text/generation?version=2023-05-29";

    const response = await axios.post<GraniteGenerationResponse>(
      apiUrl,
      buildGranitePayload(prompt.trim()),
      {
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
          Accept: "application/json",
        },
        timeout: 30000,
      }
    );

    const result = response.data.results?.[0];

    return {
      success: true,
      data: {
        model: response.data.model_id || buildGranitePayload(prompt).model_id,
        text: result?.generated_text || "",
        stopReason: result?.stop_reason || "unknown",
        tokens: {
          input: result?.input_token_count || 0,
          output: result?.generated_token_count || 0,
        },
      },
      meta: {
        timestamp: response.data.created_at || new Date().toISOString(),
      },
    };
  } catch (error) {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError<{ error?: string; errors?: Array<{ message?: string }> }>;
      const remoteMessage =
        axiosError.response?.data?.error ||
        axiosError.response?.data?.errors?.[0]?.message ||
        axiosError.message;

      throw new Error(`Granite API error: ${remoteMessage}`);
    }

    throw new Error(
      `Granite service error: ${error instanceof Error ? error.message : "Unknown error"}`
    );
  }
}

export default callGranite;

File: backend/services/granite_service.py
import os
import time
import requests
from dotenv import load_dotenv

load_dotenv()

cached_token = None
token_expires_at = 0

def get_required_env(name: str) -> str:
    value = os.getenv(name)
    if not value or not value.strip():
        raise ValueError(f"Missing required environment variable: {name}")
    return value.strip()

def get_iam_token() -> str:
    global cached_token, token_expires_at
    now = time.time() * 1000

    if cached_token and now < token_expires_at:
        return cached_token

    api_key = get_required_env("IBM_CLOUD_API_KEY")
    iam_url = os.getenv("IBM_IAM_URL", "https://iam.cloud.ibm.com/identity/token").strip()

    payload = {
        "grant_type": "urn:ibm:params:oauth:grant-type:apikey",
        "apikey": api_key
    }

    response = requests.post(
        iam_url,
        data=payload,
        headers={
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "application/json"
        },
        timeout=15
    )
    response.raise_for_status()

    data = response.json()
    cached_token = data["access_token"]
    expires_in = data.get("expires_in", 3600)
    token_expires_at = now + max(60, expires_in - 120) * 1000

    return cached_token

def call_granite(prompt: str) -> dict:
    if not prompt or not prompt.strip():
        raise ValueError("Prompt is required.")

    token = get_iam_token()
    api_url = os.getenv("IBM_GRANITE_API_URL", "https://us-south.ml.cloud.ibm.com/ml/v1/text/generation?version=2023-05-29").strip()
    model_id = os.getenv("IBM_GRANITE_MODEL_ID", "ibm/granite-3-8b-instruct").strip()
    project_id = get_required_env("IBM_WATSONX_PROJECT_ID")

    payload = {
        "model_id": model_id,
        "input": prompt,
        "parameters": {
            "decoding_method": "greedy",
            "max_new_tokens": 1000,
            "min_new_tokens": 1,
            "repetition_penalty": 1.05
        },
        "project_id": project_id
    }

    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
        "Accept": "application/json"
    }

    response = requests.post(api_url, json=payload, headers=headers, timeout=30)
    response.raise_for_status()
    
    res_data = response.json()
    results = res_data.get("results", [{}])
    generated_text = results[0].get("generated_text", "")
    stop_reason = results[0].get("stop_reason", "")
    input_tokens = results[0].get("input_token_count", 0)
    output_tokens = results[0].get("generated_token_count", 0)

    return {
        "success": True,
        "data": {
            "model": model_id,
            "text": generated_text,
            "stopReason": stop_reason,
            "tokens": {
                "input": input_tokens,
                "output": output_tokens
            }
        },
        "meta": {
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        }
    }
